src/sql/workbench/services/accountManagement/browser/accountManagementService.ts

```diff
@@ -361,8 +361,21 @@ export class AccountManagementService implements IAccountManagementService {
 	 * @return Promise to return the security token
 	 */
 	public getAccountSecurityToken(account: azdata.Account, tenant: string, resource: azdata.AzureResource): Promise<azdata.accounts.AccountSecurityToken | undefined> {
-		return this.doWithProvider(account.key.providerId, provider => {
-			return Promise.resolve(provider.provider.getAccountSecurityToken(account, tenant, resource));
+		return this.doWithProvider(account.key.providerId, async provider => {
+			let identifier = resource.toString() + '|' + tenant;
+			let originalToken: azdata.accounts.AccountSecurityToken | undefined = account.properties.tokenMap && account.properties.tokenMap[identifier] ?
+				Object.assign(account.properties.tokenMap[identifier]) : undefined;
+			let token = await provider.provider.getAccountSecurityToken(account, tenant, resource);
+			if (tenant && token) {
+				if (!account.properties.tokenMap) {
+					account.properties.tokenMap = {};
+				}
+				account.properties.tokenMap[identifier] = token;
+				if (!originalToken || originalToken.token !== token.token) {
+					await this._accountStore.addOrUpdate(account);
```

Comment:
Did you verify whether the memento is encrypted for safe storage of credentials? IIRC it's a sqlite database, or something similar, but it isn't actually encrypted in any way.

Reply:
Hmm, I was still exploring that.. if it's queryable from file system, I'll move them to secret storage.

Reply:
With secret storage, azurecore extension can own the tokens and save/retrieve when needed since we have all info so Core should not be involved. I don't think Core should do anything with it, as local cache should be integrated at the lowest level.

In case of momento I had to pass via the core layer, will clean it!

Reply:
I looked into this further, and I'm leaning towards encrypting the token json object before saving it to momento, so it would be associated with accounts directly and cleaned up when account is removed, without maintaining a separate cleaning logic.

When using secret storage it would add complexity to code when cache is force cleared/account is deleted.